File: packages/react-native-vision-camera/ios/Hybrid Objects/HybridCameraFactory.swift
///
/// VisionCamera.swift
/// VisionCamera
/// Copyright © 2025 Marc Rousavy @ Margelo
///

import AVFoundation
import Foundation
import NitroModules

class VisionCamera: HybridCameraFactorySpec {
  var cameraPermissionStatus: PermissionStatus {
    let status = AVCaptureDevice.authorizationStatus(for: .video)
    return PermissionStatus(avStatus: status)
  }

  var microphonePermissionStatus: PermissionStatus {
    let status = AVCaptureDevice.authorizationStatus(for: .audio)
    return PermissionStatus(avStatus: status)
  }

  func requestCameraPermission() -> Promise<Bool> {
    return Promise.async {
      guard Bundle.main.infoDictionary?["NSCameraUsageDescription"] != nil else {
        throw RuntimeError.error(
          withMessage:
            "Cannot request Camera Permission - Your Info.plist does not contain a `NSCameraUsageDescription` key!"
        )
      }
      return await AVCaptureDevice.requestAccess(for: .video)
    }
  }

  func requestMicrophonePermission() -> Promise<Bool> {
    return Promise.async {
      guard Bundle.main.infoDictionary?["NSMicrophoneUsageDescription"] != nil else {
        throw RuntimeError.error(
          withMessage:
            "Cannot request Microphone Permission - Your Info.plist does not contain a `NSMicrophoneUsageDescription` key!"
        )
      }
      return await AVCaptureDevice.requestAccess(for: .audio)
    }
  }

  var supportsMultiCamSessions: Bool {
    return AVCaptureMultiCamSession.isMultiCamSupported
  }

  func createCameraSession(enableMultiCam: Bool) -> Promise<any HybridCameraSessionSpec> {
    return Promise.async {
      if enableMultiCam {
        guard self.supportsMultiCamSessions else {
          throw RuntimeError.error(
            withMessage:
              "Failed to create CameraSession - this device does not support multi-cam session! (See .supportsMultiCamSession)"
          )
        }
      }
      return HybridCameraSession(enableMultiCam: enableMultiCam)
    }
  }

  func createDeviceFactory() -> Promise<any HybridCameraDeviceFactorySpec> {
    return Promise.async {
      return HybridCameraDeviceFactory()
    }
  }

  func createPhotoOutput(options: PhotoOutputOptions) throws -> any HybridCameraPhotoOutputSpec {
    return try HybridCameraPhotoOutput(options: options)
  }

  func createVideoOutput(options: VideoOutputOptions) -> any HybridCameraVideoOutputSpec {
    if options.enablePersistentRecorder == true {
      // On iOS, the `AVCaptureMovieFileOutput` does not allow flipping the Camera while
      // recording, which is a feature we call "enablePersistentRecorder".
      // So if the user plans on using persistent recordings, we need to create a custom
      // video output that uses `AVCaptureVideoDataOutput` + `AVAssetWriter` instead.
      return HybridCameraVideoFrameOutput(options: options)
    } else {
      return HybridCameraVideoOutput(options: options)
    }
  }

  func createFrameOutput(options: FrameOutputOptions) -> any HybridCameraFrameOutputSpec {
    return HybridCameraFrameOutput(options: options)
  }

  func createDepthFrameOutput(options: DepthFrameOutputOptions)
    -> any HybridCameraDepthFrameOutputSpec
  {
    return HybridCameraDepthFrameOutput(options: options)
  }

  func createObjectOutput(options: ObjectOutputOptions) -> any HybridCameraObjectOutputSpec {
    return HybridCameraObjectOutput(options: options)
  }

  func createPreviewOutput() -> any HybridCameraPreviewOutputSpec {
    return HybridCameraPreviewOutput()
  }

  func createOutputSynchronizer(outputs: [any HybridCameraOutputSpec]) throws
    -> any HybridCameraOutputSynchronizerSpec
  {
    return try HybridCameraOutputSynchronizer(outputs: outputs)
  }

  func createZoomGestureController() -> any HybridZoomGestureControllerSpec {
    return HybridZoomGestureController()
  }

  func createTapToFocusGestureController() -> any HybridTapToFocusGestureControllerSpec {
    return HybridTapToFocusGestureController()
  }

  func createOrientationManager(orientationSource: OrientationSource)
    -> any HybridOrientationManagerSpec
  {
    switch orientationSource {
    case .interface:
      return HybridInterfaceOrientationManager()
    case .device:
      return HybridDeviceOrientationManager()
    }
  }

  func createFrameRenderer() -> any HybridFrameRendererSpec {
    return HybridFrameRenderer()
  }

  func createNormalizedMeteringPoint(x: Double, y: Double, size: Double?)
    -> any HybridMeteringPointSpec
  {
    return HybridMeteringPoint(normalizedX: x, normalizedY: y, normalizedSize: size ?? 0.1)
  }

  func resolveConstraints(
    device: any HybridCameraDeviceSpec,
    outputConfigurations: [CameraOutputConfiguration],
    constraints: [Constraint],
    requiresMultiCam: Bool?
  ) -> Promise<any HybridCameraSessionConfigSpec> {
    return Promise.parallel {
      let outputs = outputConfigurations.map { $0.output }
      let resolvedConstraints = try ConstraintResolver.resolveConstraints(
        for: device,
        constraints: constraints,
        outputs: outputs,
        isMultiCam: requiresMultiCam == true)
      return HybridCameraSessionConfig(
        negotiatedFormat: resolvedConstraints.negotiatedFormat,
        enabledConstraints: resolvedConstraints.enabledConstraints)
    }
  }
}

File: packages/react-native-vision-camera/android/src/main/java/com/margelo/nitro/camera/HybridCameraFactory.kt
package com.margelo.nitro.camera

import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import androidx.annotation.Keep
import androidx.camera.core.SurfaceOrientedMeteringPointFactory
import androidx.camera.lifecycle.ProcessCameraProvider
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.bridge.ReactApplicationContext
import com.margelo.nitro.NitroModules
import com.margelo.nitro.camera.extensions.await
import com.margelo.nitro.camera.extensions.getPermissionStatus
import com.margelo.nitro.camera.extensions.requestPermission
import com.margelo.nitro.camera.hybrids.HybridCameraSession
import com.margelo.nitro.camera.hybrids.HybridCameraSessionConfig
import com.margelo.nitro.camera.hybrids.HybridFrameRenderer
import com.margelo.nitro.camera.hybrids.gestures.HybridTapToFocusGestureController
import com.margelo.nitro.camera.hybrids.gestures.HybridZoomGestureController
import com.margelo.nitro.camera.hybrids.metering.HybridMeteringPoint
import com.margelo.nitro.camera.hybrids.orientation.HybridDeviceOrientationManager
import com.margelo.nitro.camera.hybrids.orientation.HybridInterfaceOrientationManager
import com.margelo.nitro.camera.hybrids.outputs.HybridDepthFrameOutput
import com.margelo.nitro.camera.hybrids.outputs.HybridFrameOutput
import com.margelo.nitro.camera.hybrids.outputs.HybridPhotoOutput
import com.margelo.nitro.camera.hybrids.outputs.HybridPreviewOutput
import com.margelo.nitro.camera.hybrids.outputs.HybridVideoOutput
import com.margelo.nitro.camera.public.NativeCameraDevice
import com.margelo.nitro.camera.session.ConstraintResolver
import com.margelo.nitro.core.Promise

@Suppress("unused")
@Keep
@DoNotStrip
class VisionCamera : HybridCameraFactorySpec() {
  val context: ReactApplicationContext
    get() = NitroModules.applicationContext ?: throw Error("No ApplicationContext set!")
  val cameraProviderFuture = ProcessCameraProvider.getInstance(context)

  override val cameraPermissionStatus: PermissionStatus
    get() = context.getPermissionStatus(Manifest.permission.CAMERA)
  override val microphonePermissionStatus: PermissionStatus
    get() = context.getPermissionStatus(Manifest.permission.RECORD_AUDIO)
  override val supportsMultiCamSessions: Boolean
    @SuppressLint("InlinedApi")
    get() = context.packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_CONCURRENT)
  private val normalizedMeteringPointFactory: SurfaceOrientedMeteringPointFactory by lazy {
    SurfaceOrientedMeteringPointFactory(1f, 1f)
  }

  override fun requestCameraPermission(): Promise<Boolean> {
    return Promise.async {
      return@async context.requestPermission(Manifest.permission.CAMERA)
    }
  }

  override fun requestMicrophonePermission(): Promise<Boolean> {
    return Promise.async {
      return@async context.requestPermission(Manifest.permission.RECORD_AUDIO)
    }
  }

  override fun createCameraSession(enableMultiCam: Boolean): Promise<HybridCameraSessionSpec> {
    return Promise.async {
      val cameraProvider = cameraProviderFuture.await()
      return@async HybridCameraSession(cameraProvider)
    }
  }

  override fun resolveConstraints(
    device: HybridCameraDeviceSpec,
    outputConfigurations: Array<CameraOutputConfiguration>,
    constraints: Array<Constraint>,
    requiresMultiCam: Boolean?,
  ): Promise<HybridCameraSessionConfigSpec> {
    return Promise.async {
      val device =
        device as? NativeCameraDevice
          ?: throw Error("The given `device` was not of type `NativeCameraDevice`!")
      val config = ConstraintResolver.resolveConstraints(device.cameraInfo, outputConfigurations, constraints)
      return@async HybridCameraSessionConfig(device.cameraInfo, config.sessionConfig, config.resolvedConfig)
    }
  }

  override fun createDeviceFactory(): Promise<HybridCameraDeviceFactorySpec> {
    return Promise.async {
      val cameraProvider = cameraProviderFuture.await()
      return@async HybridCameraDeviceFactory(cameraProvider)
    }
  }

  override fun createPhotoOutput(options: PhotoOutputOptions): HybridCameraPhotoOutputSpec {
    return HybridPhotoOutput(options)
  }

  override fun createVideoOutput(options: VideoOutputOptions): HybridCameraVideoOutputSpec {
    return HybridVideoOutput(options)
  }

  override fun createFrameOutput(options: FrameOutputOptions): HybridCameraFrameOutputSpec {
    return HybridFrameOutput(options)
  }

  override fun createDepthFrameOutput(options: DepthFrameOutputOptions): HybridCameraDepthFrameOutputSpec {
    return HybridDepthFrameOutput(options)
  }

  override fun createPreviewOutput(): HybridCameraPreviewOutputSpec {
    return HybridPreviewOutput()
  }

  override fun createObjectOutput(options: ObjectOutputOptions): HybridCameraObjectOutputSpec {
    throw Error("CameraObjectOutput is not available on Android!")
  }

  override fun createOutputSynchronizer(outputs: Array<HybridCameraOutputSpec>): HybridCameraOutputSynchronizerSpec {
    throw Error("OutputSynchronizer is not available on Android!")
  }

  override fun createZoomGestureController(): HybridZoomGestureControllerSpec {
    return HybridZoomGestureController()
  }

  override fun createTapToFocusGestureController(): HybridTapToFocusGestureControllerSpec {
    return HybridTapToFocusGestureController()
  }

  override fun createOrientationManager(orientationSource: OrientationSource): HybridOrientationManagerSpec {
    return when (orientationSource) {
      OrientationSource.INTERFACE -> HybridInterfaceOrientationManager()
      OrientationSource.DEVICE -> HybridDeviceOrientationManager()
    }
  }

  override fun createFrameRenderer(): HybridFrameRendererSpec {
    return HybridFrameRenderer()
  }

  override fun createNormalizedMeteringPoint(
    x: Double,
    y: Double,
    size: Double?,
  ): HybridMeteringPointSpec {
    val meteringPoint =
      if (size != null) {
        normalizedMeteringPointFactory.createPoint(x.toFloat(), y.toFloat(), size.toFloat())
      } else {
        normalizedMeteringPointFactory.createPoint(x.toFloat(), y.toFloat())
      }
    return HybridMeteringPoint(
      x,
      y,
      size,
      meteringPoint,
    )
  }
}
